SQL/proc_descriptions/procedure_assign_average_speed_to_all_segments_in_area.md

```diff
@@ -0,0 +1,110 @@
+        - CTE `target_ways` executes function `get_ways_in_target_area`, which returns table used by `node_segments` CTE. Note: Did not figure out the role it demonstrates to the procedure.
+        - CTE `node_segments` returns a list of records with columns `from_id`, `to_id`, `geom`. `geom` column is not used afterwards, `from_id` and `to_id` are gathered from `nodes_ways` table, where the latter is joined by some specific filtering. Note: more on QA
+    - Final insertion is made by a list, which is created by Cross Join aka Cartesian product of `node_segments` and `average_speed`, which actually leaves dimensionality to (`count(*) FROM node_segments` X 1), because of one line from `average_speed`. Isertion into `nodes_ways_speeds` with `quality`=5 is executed on the end.
+- __Simplifying it all__, it takes records from `nodes_ways` table, adds some additional data from the same table, then adds additional columns such as `average_speed` & `average_st_dev` and inserts to `nodes_ways_speeds` with `quality`=5.
+
+## Todo list
+This part is about suggestions of what could be tested. On the end of every point you can see _Status_, which shows what status is on every point, if approven corresponding `approved` should be shown.
+- Testing __procedure__:
+    - test that after execution of this procedure, some records were added to `nodes_ways_speeds` with corresponding values (which would be hardcoded to the assertion). Status: `on review`
+    - test that after double execution with the same args, records wouldn't be added second time. Status: `on review`
+    - test that after inserting invalid args (either of them), the procedure wouldn't modify table, but throw an exception (we may need to modify this procedure to throw an exception). Status: `on review`
+    - test `get_ways_in_target_area()` as an individual function (I mean to test it independently from testing this procedure). Status: `on review`
+- Modification of __procedure__:
+    - add throwing an exception if args are invalid on the start of the procedure (invalid in this context could mean for example, that corresponding `area` does not exist, or some records referring to this area exist in one table, but do not exist in another used in this procedure). Status: `on review`
+    - add throwing an exception if `nodes_ways_speeds` table before execution contains no records. Status: `on review`
+
+## QA
+- Q: Not sure what __segment__ in this particular context means. I see that we work with the so-called __ways__, does segment refer to that. A: -
+- Q: do not really think that it will matter, but anyway we seem to use column `quality` from `nodes_ways_speeds` in this procedure. We gather info on average speed/std deviation from records with quality 1 or 2, and then inserting in the same table with quality 5. A: -
```

Comment:
The quality should be derived from the quality of the sources (worst source among all segments). But it does not matter so far as we do not use the quality anywhere...